runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -242,41 +253,20 @@ public int compareTo(CorfuQueueRecord<? extends E> o) {
             );
         }
 
-        // Bind the iteration order to a snapshot of the Queue using a transaction.
-        long snapshotVersion;
-        boolean startedNewTransaction = false;
-        if (TransactionalContext.isInTransaction()) {
-            snapshotVersion = TransactionalContext.getCurrentContext()
-                    .getSnapshotTimestamp().getSequence();
-        } else {
-            runtime.getObjectsView().TXBuild().type(TransactionType.WRITE_AFTER_WRITE)
-                    .build()
-                    .begin();
-            snapshotVersion = TransactionalContext.getCurrentContext()
-                    .getSnapshotTimestamp().getSequence();
-            startedNewTransaction = true;
-        }
         List<CorfuQueueRecord<E>> copy = new ArrayList<>(
                 Math.min(corfuTable.size(), maxEntries)
         );
 
+    Comparator<Map.Entry<Long, CorfuQueueRecord<E>>> recordIdComparator = (r1, r2) ->
+                r1.getValue().getRecordId().compareTo(r2.getValue().recordId);
         int index = 0;
-        for (Long entryId : corfuTable.keySet().stream()
-                .filter(e -> e > entriesAfter)
-                .sorted().collect(Collectors.toList())) {
+        for (Map.Entry<Long, CorfuQueueRecord<E>> entry : corfuTable.entryStream()
+                .filter(e -> e.getValue().getRecordId().compareTo(entriesAfter) > 0)
+                .sorted(recordIdComparator).collect(Collectors.toList())) {
             if (++index >= maxEntries) {
```

Comment:
You can put a `.limit(maxEntries)` after `.filter()` to avoid collecting and sorting all entries and ++index. This should be more efficient.

Reply:
done. thanks